packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2019 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import webpack from 'webpack';
+import {SyncWaterfallHook} from 'tapable';
+
+const {Template} = webpack;
+
+/**
+ * Webpack Plugin to batch dynamic chunk loading requests.
+ *
+ * <p>This plugin is compatible with both Webpack 4 and 5 APIs. It works by
+ * overriding parts of the webpack runtime. Instead of immediately generating
+ * script tags for each chunk, it uses a Promise to queue a task in the
+ * microtask queue that creates a single script tag for all chunks requested
+ * within the same synchronous code flow.
+ *
+ * Because of limitations in webpack's internal code structure, we always need
+ * to return some DOM node for each chunk. We return the actual script tag
+ * for the first call in a batch and comment nodes for subsequent calls.
+ *
+ * Most of the generated code in the Tempate is 1:1 the code generated by
+ * webpack's default implementation, divergence is clearly marked with DIFF.
+ */
+export default class DynamicBundlePlugin {
+  /**
+   * @param buildBatchEndpoint Helper to construct an asset URL from a serving
+   *                           prefix and a list of chunks ids.
+   */
+  constructor(private buildBatchEndpoint: (prefix: string, ids: number[]) => string) {
+  }
+
+  apply(compiler: webpack.Compiler) {
+    compiler.hooks.compilation.tap(DynamicBundlePlugin.name, (compilation) => {
```

Comment:
nit: is the styleguide to use (foo) => {} or foo => {}?

Reply:
Can't find a clear note on this in our internal styleguide but once the type annotation is added, the non-parenthesis style is no longer valid syntax.